currencies/rate_converter.go

```diff
@@ -0,0 +1,205 @@
+	ticker                *time.Ticker
+	fetchingInterval      time.Duration
+	retryFetchingInterval time.Duration
+	syncSourceURL         string
+	rates                 atomic.Value // Should only hold Rates struct
+	lastUpdated           atomic.Value // Should only hold time.Time
+}
+
+// NewRateConverter returns a new RateConverter
+func NewRateConverter(
+	httpClient httpClient,
+	syncSourceURL string,
+	fetchingInterval time.Duration,
+	retryFetchingInterval time.Duration,
+) *RateConverter {
+	return NewRateConverterWithNotifier(
+		httpClient,
+		syncSourceURL,
+		fetchingInterval,
+		retryFetchingInterval,
+		nil, // no notifier channel specified, won't send any notifications
+	)
+}
+
+// NewRateConverterWithNotifier returns a new RateConverter
+// it allow to pass an update chan in which the number of ticks will be passed after each tick
+// allowing clients to listen on updates
+// Do not use this method
+func NewRateConverterWithNotifier(
+	httpClient httpClient,
+	syncSourceURL string,
+	fetchingInterval time.Duration,
+	retryFetchingInterval time.Duration,
+	updateNotifier chan<- int,
+) *RateConverter {
+	rc := &RateConverter{
+		httpClient:            httpClient,
+		updates:               make(chan *Rates),
+		done:                  make(chan bool),
+		updateNotifier:        updateNotifier,
+		fetchingInterval:      fetchingInterval,
+		retryFetchingInterval: retryFetchingInterval,
+		ticker:                nil,
+		syncSourceURL:         syncSourceURL,
+		rates:                 atomic.Value{},
+		lastUpdated:           atomic.Value{},
+	}
+
+	// In case host do not want to support currency lookup
+	// we just stop here and do nothing
+	if rc.fetchingInterval == time.Duration(0) {
+		return rc
+	}
+
+	rc.Update()                // Make sure to populate data before to return the converter
+	rc.startPeriodicFetching() // Start periodic ticking
+
+	return rc
+}
+
+// fetch allows to retrieve the currencies rates from the syncSourceURL provided
+func (rc *RateConverter) fetch() (*Rates, error) {
+	request, err := http.NewRequest("GET", rc.syncSourceURL, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	defer cancel()
+	request = request.WithContext(ctx)
+
+	response, err := rc.httpClient.Do(request)
+	if err != nil {
+		return nil, err
+	}
+
+	defer response.Body.Close()
+
+	bytesJSON, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	updatedRates := &Rates{}
+	err = json.Unmarshal(bytesJSON, &updatedRates)
+	if err != nil {
+		return nil, err
+	}
+
+	return updatedRates, err
+}
+
+// updateRates allows to update the internal currencies rates map with the one provided
+func (rc *RateConverter) updateRates(updatedRates *Rates) {
+	rc.rates.Store(updatedRates)
+	rc.lastUpdated.Store(time.Now())
+}
+
+// Update updates the internal currencies rates from remote sources
+func (rc *RateConverter) Update() (*Rates, error) {
```

Comment:
Is the `*Rates` return value used for anything? For that matter do you ever try checking the `error` state when you call it?

Reply:
I don't use `*Rates` but in tests. I can totally remove it you are right. 

Regarding error, I added error logging / check in `Update()` function directly. I also still think returning the error is not bad, it's used to test the function and it also allows to be able to know if there was an error or not in the calling function (clients might want to handle the error or do something about it).